### PR TITLE
3.x: Backport of "Make metadata cache key configurable"

### PR DESCRIPTION
**Pull request by bancer**
Issue https://github.com/cakephp/cakephp/issues/13471

https://github.com/cakephp/cakephp/pull/13472 4.x backport to 3.x

[edited by dereuromark - please always add these things into description instead of title]

### CODE REVIEW

File: src/Database/Connection.php [hunk 1]
```diff
@@ -107,7 +107,14 @@ class Connection implements ConnectionInterface
     /**
      * Constructor.
      *
-     * @param array $config configuration for connecting to database
+     * @param array $config Configuration array.
+     * ### Available options:
```
Comment (dereuromark) [inline src/Database/Connection.php]: I would prefer those going on top before params start, as we usually do in other places.

File: src/Database/Schema/CachedCollection.php [hunk 2]
```diff
@@ -77,7 +77,12 @@ public function describe($name, array $options = [])
      */
     public function cacheKey($name)
     {
-        return $this->_connection->configName() . '_' . $name;
+        $cachePrefix = $this->_connection->configName();
+        $config = $this->_connection->config();
+        if (isset($config['cacheKeyPrefix'])) {
+            $cachePrefix = $config['cacheKeyPrefix'];
+        }
+        return $cachePrefix . '_' . $name;
```
Comment (dereuromark) [inline src/Database/Schema/CachedCollection.php]: I would consider a "true" prefix to include the `_` here, no magic internal additions IMO. Or is this how it works in 4.x now as well?

Comment (bancer) [reply src/Database/Schema/CachedCollection.php]: I do not understand that. It is `return $this->prefix . '_' . $name;` in https://github.com/cakephp/cakephp/blob/4.x/src/Database/Schema/CachedCollection.php#L98.

Comment (dereuromark) [reply src/Database/Schema/CachedCollection.php]: I see.
I would like to get feedback here from @ADmad @markstory etc on if we can change 4.x here also to make prefixing behave as in other cases - and how intuitively expected (to not magically append another glue char.

Comment (ADmad) [reply src/Database/Schema/CachedCollection.php]: I am fine with not auto adding `_` if prefix has been explicitly specified if someone wants to update 4.x.

Comment (dereuromark) [reply src/Database/Schema/CachedCollection.php]: But I see also
- https://github.com/cakephp/cakephp/blob/69643f58dd110a4c5cce7ff867707de176c64bc4/src/View/Helper/IdGeneratorTrait.php#L98

which is in line with this change however. Even though that is for HTML DOM - and not as file, path or cache prefix.

Comment (dereuromark) [reply src/Database/Schema/CachedCollection.php]: The doc block on 4.x states

 > Typically the connection name.

As such it is probably fine to keep it as is (and auto append the underscore).

Comment (markstory) [reply src/Database/Schema/CachedCollection.php]: I don't see a problem with the underscore. It is a consistent input to the cache key and makes keys more readable should you need to inspect them.

**Comment by ADmad** [comment]
Even though there's no API change here adding feature flags in bugfix releases makes it difficult to keep track of since when a feature is available.

**Comment by dereuromark** [comment]
Putting it into 3.next then sounds like a fair compromise

**Comment by dereuromark** [comment]
This also needs a documentation update for the migration guide and specific docs on the topic.

**Pull request merged**